### PR TITLE
nodejs: use ninja for build

### PR DESCRIPTION
**Pull request by tie**
## Description of changes

This PR contains changes to build Node.js without cross-compilation (see https://github.com/NixOS/nixpkgs/pull/220204#issuecomment-2229544988) and use ninja for build.

Note that we can’t use `--cross-compiling` with Ninja without https://github.com/nodejs/gyp-next/pull/185, but we don’t have to because we use an emulator instead. This is based on Buildroot’s [0001-add-qemu-wrapper-support.patch](https://gitlab.com/buildroot.org/buildroot/-/blob/c1d5eada4d4db9eeaa1c44dd1dea95a67c8a70ca/package/nodejs/nodejs-src/0001-add-qemu-wrapper-support.patch) patch, but re-implemented in a manner that hopefully can be upstreamed (see also https://github.com/nodejs/node/pull/53899).

__Supersedes__
- #220204 (thanks @Cynerd for working on this!)
- #317109

__Fixes__
- #321847
- #319036
- #326150

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] aarch64-linux → i686-linux (pkgsCross.gnu32)
  - [x] aarch64-linux → armv6l-linux (pkgsCross.raspberryPi but
    `gcc.arch = "armv6kz"` and `gcc.fpu = "vfpv2"`)
  - [x] aarch64-linux → armv7l-linux (pkgsCross.armv7l-hf-multiplatform)
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/web/nodejs/nodejs.nix [hunk 1]
```diff
@@ -229,20 +343,9 @@ let
...
-        # llvm-ar supports response files, so take advantage of it if it’s available.
-        if [ "$(basename $(readlink -f $(command -v ar)))" = "llvm-ar" ]; then
-          ar -cqs $libv8/lib/libv8.a @files
-        else
-          cat files | while read -r file; do
-            ar -cqS $libv8/lib/libv8.a $file
-          done
-        fi
-      ''}
+      $AR -cqs $libv8/lib/libv8.a @files
```
Comment (tie) [inline pkgs/development/web/nodejs/nodejs.nix:348]: Note: before this change, we had fallback for `ar` that does not support response files. As far as I know, that is only the case for Darwin’s cctools, but stdenv already uses LLVM for most of the tools on Darwin.
https://github.com/NixOS/nixpkgs/blob/034f3eec9b56e2fa22ee77f67403310c202ed398/pkgs/os-specific/darwin/binutils/default.nix#L117

That is, I think it makes sense to assume that `ar` can handle response files (unless there is some edge case I’m missing).

Comment (tie) [reply pkgs/development/web/nodejs/nodejs.nix:348]: In the future, ideally, we should update v8 package and avoid using nodejs as v8 library provider.

**Review by aduh95 — APPROVED** [review]
Just tested that on a aarch64-darwin machine:
```console
$ nixpkgs-review pr   --checkout commit   --package nodejs_18   --package nodejs_20   --package nodejs_22   327653

6 packages built:
nodejs_18 nodejs_18.libv8 nodejs_20 nodejs_20.libv8 nodejs_22 nodejs_22.libv8
```

File: pkgs/development/web/nodejs/bypass-darwin-xcrun-node16.patch [hunk 2]
```diff
@@ -1,7 +1,42 @@
...
++++ b/tools/gyp/pylib/gyp/xcode_emulation.py
+@@ -522,7 +522,13 @@ class XcodeSettings:
+         # Since the CLT has no SDK paths anyway, returning None is the
+         # most sensible route and should still do the right thing.
+         try:
+-            return GetStdoutQuiet(["xcrun", "--sdk", sdk, infoitem])
++            #return GetStdoutQuiet(["xcrun", "--sdk", sdk, infoitem])
++            return {
++                "--show-sdk-platform-path": "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform",
++                "--show-sdk-path": "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk",
```
Comment (emilazy) [inline pkgs/development/web/nodejs/bypass-darwin-xcrun-node16.patch:13]: How does this work with the build sandbox?

File: pkgs/development/web/nodejs/nodejs.nix
Comment (emilazy) [inline pkgs/development/web/nodejs/nodejs.nix:1]: Could you do the reformatting in a separate commit?

**Review by emilazy — CHANGES_REQUESTED** [review]
Do you know if this fixes the weird double‐compilation test race condition, or is that still yet to be diagnosed?

File: pkgs/development/web/nodejs/bypass-darwin-xcrun-node16.patch (hunk 2 above)
Comment (tie) [reply pkgs/development/web/nodejs/bypass-darwin-xcrun-node16.patch:13]: Node.js has two copies of GYP and this change just extends the existing patch to the second GYP copy. This works fine *with* sandbox, but I'm not really sure how that's supposed to work *without* sandbox since technically nothing stops the build from accessing these paths.

Anyway, this allows us to drop xcbuild from Node.js native build inputs with minor copy-paste modification to an existing patch. I'm leaving refactoring or documenting this behavior to my future self (e.g. if we add split dev output, we can use fake xcbuild in propagatedNativeBuildInputs instead of patching GYP).

**Comment by tie** [comment]
>Do you know if this fixes the weird double‐compilation test race condition, or is that still yet to be diagnosed?

Not yet, but I think Node.js had the same issue that they fixed by limiting Make parallelism: https://github.com/nodejs/node/issues/22006#issuecomment-408697797

We are not using `test-ci` target since _IIRC_ that never worked in Nix sandbox and requires network access, but we should be able to apply the Nix equivalent of the upstream fix, that is, `enableParallelChecking = false`.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (tie) [reply pkgs/development/web/nodejs/nodejs.nix:1]: I’ll reformat in all nodejs package files a separate PR 👌
There are a few more code style issues beyond formatting that I’d like to address.

**Comment by tie** [comment]
Force-pushed to minimize formatting changes (see https://github.com/NixOS/nixpkgs/pull/327653#discussion_r1680724950). This causes rebuilds because of `postInstall`.

File: pkgs/development/web/nodejs/nodejs.nix [hunk 3]
```diff
@@ -16,35 +15,110 @@
...
+    else if platform.isx86_32 then
+      "ia32"
+    else if platform.isS390x then
+      "s390x"
+    else if platform.isRiscV64 then
+      "riscv64"
+    else if platform.isLoongArch64 then
+      "loong64"
+    else
+      throw "unsupported cpu ${platform.uname.processor}";
```
Comment (tie) [inline pkgs/development/web/nodejs/nodejs.nix:70]: This is not an idiomatic Nixpkgs pattern and should probably use some fallback for unknown platforms instead of throwing an error. I’m keeping it for now and will address this in subsequent code style PRs.

File: pkgs/development/web/nodejs/nodejs.nix [hunk 4]
```diff
@@ -129,7 +129,11 @@ let
 
     strictDeps = true;
 
-    env = lib.optionalAttrs (stdenv.isDarwin && stdenv.isx86_64) {
+    env = {
+      # Tell ninja to avoid ANSI sequences, otherwise we don’t see build
+      # progress in Nix logs.
+      NINJA = "TERM=dumb ninja";
```
Comment (tie) [inline pkgs/development/web/nodejs/nodejs.nix:135]: Not setting `TERM=dumb` globally because Node.js v18 needs https://github.com/nodejs/node/commit/f76b28f2cf0cde9dd9d4476b9bbd772bc48f722c backport.

**Comment by tie** [comment]
@emilazy, I’ve added a commit that should fix races in check phase.

**Comment by emilazy** [comment]
Thanks! Do you think there’s any chance of splitting off the tests into a separate derivation like you suggested? Maybe that should wait for a second PR…

I can test this on `x86_64-darwin` if it’s ready.

**Review by Cynerd — APPROVED** [review]

**Comment by tie** [comment]
>I can test this on x86_64-darwin if it’s ready.

Yes, thanks, the PR is ready. I’ve run built the package a few times on aarch64-darwin and didn’t notice any issues. Note: `--rebuild`/`--check` flag does not work for pre-v22 versions (yet?), see https://github.com/nodejs/build/issues/3043#issuecomment-2173715373 (that is, nodejs_18 and nodejs_20 builds are not reproducible).

>Do you think there’s any chance of splitting off the tests into a separate derivation like you suggested? Maybe that should wait for a second PR…

The change set of this PR is complete and I’d like to avoid adding more unrelated changes.

**Comment by jue89** [comment]
Did some basic testing with armv6. My hacky flake:

```nix
{
  inputs = {
    nixpkgs.url = "github:tie/nixpkgs/dc174b5cb8be64c168ad29bde661d23058501161";
  };

  outputs = { self, nixpkgs }: {
    packages.armv6l-linux = import nixpkgs {
      localSystem = {
        system = "x86_64-linux";
      };
      crossSystem = nixpkgs.lib.systems.examples.raspberryPi // {
        gcc = {
          arch = "armv6kz";
          fpu = "vfpv2";
        };
      };
    };

    packages.x86_64-linux = import nixpkgs {
      system = "x86_64-linux";
    };
  };
}
```
---

```
$ nix build --system armv6l-linux .#nodejs_20
$ file ./result/bin/node 
./result/bin/node: ELF 32-bit LSB executable, ARM, EABI5 version 1 (GNU/Linux), dynamically linked, interpreter /nix/store/lfv020g79z6xahbj7spn2gbdpa0q7mpy-glibc-armv6l-unknown-linux-gnueabihf-2.39-52/lib/ld-linux-armhf.so.3, for GNU/Linux 3.10.0, not stripped
$ ./result/bin/node -e "console.log(require('os').arch())"
arm
```

✅ Running node v20 using qemu looks good

---

```
$ nix build --system armv6l-linux .#nodejs_22
$ file ./result/bin/node 
result/bin/node: ELF 32-bit LSB executable, ARM, EABI5 version 1 (GNU/Linux), dynamically linked, interpreter /nix/store/lfv020g79z6xahbj7spn2gbdpa0q7mpy-glibc-armv6l-unknown-linux-gnueabihf-2.39-52/lib/ld-linux-armhf.so.3, for GNU/Linux 3.10.0, not stripped
$ ./result/bin/node -e "console.log(require('os').arch())"
arm
```

✅ Running node v22 using qemu looks good

---

I'll try to run node on a RPi1 during the weekend.

But this looks really promising! Thank you a lot for tackling this problem :-)

**Comment by tie** [comment]
Nitpick: I don’t think you don’t need `--system armv6l-linux` for the Nix invocations above.

E.g.

<details>
<summary>flake.nix</summary>

```nix
{
  inputs = {
    nixpkgs.url = "github:tie/nixpkgs/dc174b5cb8be64c168ad29bde661d23058501161";
  };

  outputs =
    inputs:
    let
      # Tip: if you only need lib but not Nixpkgs (e.g. for Nix libraries
      # distributed as flakes), you can use lib as a standalone flake.
      # {
      #   inputs.nixpkgs-lib.url = "github:NixOS/nixpkgs/nixos-unstable?dir=lib";
      #   outputs = inputs: let inherit (inputs.nixpkgs-lib) lib; in { … };
      # }
      inherit (inputs.nixpkgs) lib;

      # Tip: use flake-parts (https://flake.parts) to make parametrizing over
      # system easier using a familiar NixOS module system.
      system = "aarch64-linux";

      pkgsForRaspberryPi = import inputs.nixpkgs {
        localSystem = {
          inherit system;
        };
        crossSystem = lib.systems.examples.raspberryPi // {
          gcc = {
            arch = "armv6kz";
            fpu = "vfpv2";
          };
        };
      };
    in
    {
      packages.${system} = lib.mapAttrs' (name: value: lib.nameValuePair ("raspberry-pi-" + name) value) (
        lib.getAttrs [
          "nodejs"
          "nodejs_18"
          "nodejs_20"
          "nodejs_22"
          "nodejs_latest"
          "nodejs-slim"
          "nodejs-slim_18"
          "nodejs-slim_20"
          "nodejs-slim_22"
          "nodejs-slim_latest"
        ] pkgsForRaspberryPi
      );
    };
}
```
</details>

```
nix build .#raspberry-pi-nodejs
```

File: pkgs/development/web/nodejs/bypass-darwin-xcrun-node16.patch
Comment (SuperSandro2000) [inline pkgs/development/web/nodejs/bypass-darwin-xcrun-node16.patch:1]: Please delete the old code instead of commenting it

File: pkgs/development/web/nodejs/nodejs.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/development/web/nodejs/nodejs.nix:70]: It's fine to throw if we don't know the platform.

File: pkgs/development/web/nodejs/nodejs.nix [hunk 5]
```diff
@@ -16,35 +15,110 @@
...
+  # would shadow stdenv’s bintools and potentially break other parts of the
+  # build. The correct behavior is to use absolute paths, and there is a PR for
+  # that, see https://github.com/NixOS/nixpkgs/pull/314920. As a temporary
+  # workaround, we use only a single program we need (and that is not part of
+  # the stdenv).
+  darwin-cctools-only-libtool =
+    # Would be nice to have onlyExe builder similar to onlyBin…
+    runCommand "darwin-cctools-only-libtool" { cctools = lib.getBin buildPackages.cctools; } ''
+      mkdir -p -- "$out/bin"
+      ln -s -- "$cctools/bin/libtool" "$out/bin/libtool"
```
Comment (SuperSandro2000) [inline pkgs/development/web/nodejs/nodejs.nix:119]:
```suggestion
      mkdir -p "$out/bin"
      ln -s "$cctools/bin/libtool" "$out/bin/libtool"
```

File: pkgs/development/web/nodejs/nodejs.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/development/web/nodejs/nodejs.nix:135]: We should write that into a comment

File: pkgs/development/web/nodejs/nodejs.nix [hunk 6]
```diff
@@ -55,58 +129,89 @@ let
...
+          as that would put the paths into bin/nodejs.
+          Including pkg-config in build inputs would also have the same effect!
+
+          FIXME: the statement above is outdated, we have to include pkg-config
+          in build inputs for system-icu.
+        */
+      ]) (builtins.attrNames sharedLibDeps)
+      ++ [
+        "--with-intl=system-icu"
+        "--openssl-use-def-ca-store"
```
Comment (SuperSandro2000) [inline pkgs/development/web/nodejs/nodejs.nix:207]: Please move all none conditional things to the beginning

File: pkgs/development/web/nodejs/bypass-darwin-xcrun-node16.patch
Comment (emilazy) [reply pkgs/development/web/nodejs/bypass-darwin-xcrun-node16.patch:1]: That would be inconsistent with the part of the patch that applies to the other copy of this code, which was already in‐tree. Is it important enough to re‐do the whole patch over?

Comment (tie) [reply pkgs/development/web/nodejs/bypass-darwin-xcrun-node16.patch:1]: See https://github.com/NixOS/nixpkgs/pull/327653#discussion_r1681159096
I don’t think it’s worth redoing this patch now.

File: pkgs/development/web/nodejs/nodejs.nix (hunk 4 above)
Comment (emilazy) [reply pkgs/development/web/nodejs/nodejs.nix:135]: Hmm, this is unfortunate. I think Nix build environments have enough of a PTY to run these tests, but we’re using `TERM=dumb` to get Ninja to give us more detailed output. It’d be nice if we could still run those tests, but it’s probably not that important.

File: pkgs/development/web/nodejs/nodejs.nix [hunk 7]
```diff
@@ -207,12 +311,12 @@ let
     postInstall = ''
       HOST_PATH=$out/bin patchShebangs --host $out
 
-      ${lib.optionalString (stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
-        $out/bin/${self.meta.mainProgram} --completion-bash > ${self.meta.mainProgram}.bash
-        installShellCompletion ${self.meta.mainProgram}.bash
+      ${lib.optionalString canExecute ''
+        $out/bin/node --completion-bash > node.bash
+        installShellCompletion node.bash
```
[the remark this answers is not in the document]
Comment (tie) [reply pkgs/development/web/nodejs/nodejs.nix:316]: I don’t agree with your suggestion, e.g. consider
```
% bash -c 'set -e -u -o pipefail; cat <(false); echo exit code: $?'
exit code: 0
```

That is, we don’t want to continue execution if completion generation fails.

File: pkgs/development/web/nodejs/nodejs.nix (hunk 4 above)
Comment (tie) [reply pkgs/development/web/nodejs/nodejs.nix:135]: These tests run fine _unless_ `TERM=dumb` is set 😅

Comment (emilazy) [reply pkgs/development/web/nodejs/nodejs.nix:138]: Yeah. We could override `TERM` for just those tests, but it’d be ugly. I’m guessing Ninja has probably already been approached about an alternate way to toggle non‐fancy build output and maybe rejected it. Oh well.

File: pkgs/development/web/nodejs/nodejs.nix (hunk 1 above)
Comment (reckenrode) [reply pkgs/development/web/nodejs/nodejs.nix:340]: You could set `env.AR` to `"${lib.getBin llvm}/bin/llvm-ar"` to play it really safe, but that’s not necessary.

Darwin bintools has been using `llvm-ar` since 23.11. Going forward, more tools will be switched to their LLVM versions as compatibility and stability improve. None are expected to switch back to their cctools versions.

**Review by jue89 — APPROVED** [review]
I tried building a minimal armv6l NixOS image for Raspberry Pi 1 including your NodeJS fixes and failed producing a working image. Since I'm busy for the next 2 weeks, I don't think I'll have the time to finish testing on armv6l soon.

But I think working NodeJS under Qemu is already a huge improvement compared to the current situation on master. Won't block - ACK from my side.

File: pkgs/development/web/nodejs/nodejs.nix (hunk 7 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/development/web/nodejs/nodejs.nix:317]: installShellCompletion fails if the generated completion is empty.

[the remark this answers is not in the document]
Comment (tie) [reply pkgs/development/web/nodejs/nodejs.nix:317]: @SuperSandro2000, yes, but it doesn’t fail if the completion generation fails halfway through (i.e. writes at least one byte to the output). We shouldn’t be ignoring non-zero exit codes without a reason. I’m fine with your suggestion _if_ there was a way to tell Bash to actually errexit in this case, but alas there is none I’m aware of.

Even correctness and reproducibility aside, installShellCompletion has the same logic internally: write to file, copy this file to the shell-specific location. Using a named pipe just adds an unnecessary level of indirection.

https://github.com/NixOS/nixpkgs/blob/e4eb1c74f532917351fdfd6520f0fbb017159509/pkgs/build-support/setup-hooks/install-shell-files.sh#L215-L226

**Comment by agj** [comment]
Coming in from [here](https://github.com/NixOS/nixpkgs/issues/326150#issuecomment-2245404475). Trying to build Node 20 using this PR's branch in macOS Monterey I got a build taking 14 hours and with [this 34k+ lines log](https://gist.github.com/agj/b386d3820a3fa8e3eb643dd49581f41d). But here's the last few lines:

```
Failed tests:
out/Release/node /private/tmp/nix-build-nodejs-20.15.1.drv-0/node-v20.15.1/test/parallel/test-fs-readv.js
out/Release/node /private/tmp/nix-build-nodejs-20.15.1.drv-0/node-v20.15.1/test/parallel/test-fs-readv-sync.js
make: *** [Makefile:553: test-ci-js] Error 1
``

**Comment by tie** [comment]
@agj, thanks! Do you have `sandbox = true` in you `nix.conf`? Just asking so that I can reproduce the failure.
<details>

```
not ok 928 parallel/test-fs-readv
  ---
  duration_ms: 249.27500
  severity: fail
  exitcode: 1
  stack: |-
    node:assert:126
      throw new AssertionError(obj);
      ^
    
    AssertionError [ERR_ASSERTION]: Expected values to be strictly deep-equal:
    
    54 !== 109
    
        at /private/tmp/nix-build-nodejs-20.15.1.drv-0/node-v20.15.1/test/parallel/test-fs-readv.js:29:12
        at /private/tmp/nix-build-nodejs-20.15.1.drv-0/node-v20.15.1/test/common/index.js:430:17
        at /private/tmp/nix-build-nodejs-20.15.1.drv-0/node-v20.15.1/test/common/index.js:465:15
        at FSReqCallback.wrapper [as oncomplete] (node:fs:768:5) {
      generatedMessage: true,
      code: 'ERR_ASSERTION',
      actual: 54,
      expected: 109,
      operator: 'deepStrictEqual'
    }
    
    Node.js v20.15.1
  ...
ok 929 parallel/test-fs-readv-promises
  ---
  duration_ms: 266.46800
  ...
not ok 930 parallel/test-fs-readv-sync
  ---
  duration_ms: 189.69900
  severity: fail
  exitcode: 1
  stack: |-
    node:assert:126
      throw new AssertionError(obj);
      ^
    
    AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
    
    54 !== 109
    
        at Object.<anonymous> (/private/tmp/nix-build-nodejs-20.15.1.drv-0/node-v20.15.1/test/parallel/test-fs-readv-sync.js:37:10)
        at Module._compile (node:internal/modules/cjs/loader:1358:14)
        at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
        at Module.load (node:internal/modules/cjs/loader:1208:32)
        at Module._load (node:internal/modules/cjs/loader:1024:12)
        at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:174:12)
        at node:internal/main/run_main_module:28:49 {
      generatedMessage: true,
      code: 'ERR_ASSERTION',
      actual: 54,
      expected: 109,
      operator: 'strictEqual'
    }
    
    Node.js v20.15.1
  ...
```
</details>

**Comment by agj** [comment]
> @agj, thanks! Do you have `sandbox = true` in you `nix.conf`? Just asking so that I can reproduce the failure.

Ah, no, it's set to `false`. I guess it's better the other way around for these purposes? I'm still a Nix novice.

**Comment by tie** [comment]
>Ah, no, it's set to false. I guess it's better the other way around for these purposes?

Uh, it depends… This option doesn’t affect derivations when changed to a different setting (i.e. doesn’t trigger rebuilds), but does make significant changes to the build environment (e.g. forbids access to files outside of the Nix store). Not sure if there were any discussions on changing this behavior though. This is just one a few impurities in Nix builds.

Some packages don’t build under Nix sandbox on macOS, but I don’t know the exact numbers. I’ve only had issues with a single package—`openldap` needs network for tests, and that can be easily fixed with `doCheck = false`.

For cache.nixos.org, macOS packages are built without sandbox (`sandbox = false`, or maybe `relaxed`, not sure). As long as the store path can be substituted, sandbox doesn’t really matter (because it doesn’t affect derivation hash). If you are _contributing_ to Nixpkgs, I’d say it’s best to set `sandbox = true` (though note that I’m no authority on this).

I had it set to `sandbox = true` but perhaps there are some test failures in Node.js only with sandbox disabled? Or it could be x86_64-specific. I’ll look into this on the weekend.

**Comment by emilazy** [comment]
Do we think this is ready? It’s not clear to me what the outcome of the previous exchange is and I don’t want to start a flurry of builds to test if they’ll be premature :)

**Comment by agj** [comment]
I tried to build again, but no luck, although I get [a different output](https://gist.github.com/agj/eece6e40b61293b73ddc8fba5da5aaaf). Looks like a libuv build issue.

```
[0;31mFAIL[m: test/run-tests
[0;31m======================================================[m
[0;31m1 of 1 test failed[m
[0;31mPlease report to https://github.com/libuv/libuv/issues[m
[0;31m======================================================[m
make[1]: *** [Makefile:5647: check-TESTS] Error 1
make[1]: Leaving directory '/private/tmp/nix-build-libuv-1.48.0.drv-0/source'
make: *** [Makefile:5912: check-am] Error 2
```

**Comment by tie** [comment]
@emilazy, it’s ready for review and tests, hopefully ofborg-eval error will go away after force push (it’s not caused by changes in Node.js package). Last time I’ve tested this PR on `aarch64-linux`, `x86_64-linux` and `aarch64-darwin` (with sandbox enabled). I haven’t tested _without_ sandbox on macOS and also `x86_64-darwin`, @agj reports that there are some issues in the check phase with the latter.

**Comment by emilazy** [comment]
The Hydra macOS builders have the sandbox turned off, so if we want to stop the Node.js packages being so flaky on them we’ll need to figure out what’s going on there.

**Comment by tie** [comment]
Hm, I can’t reproduce build failure on `aarch64-darwin` with and without Nix sandbox. Perhaps it’s really something specific to `x86_64-darwin`.

**Comment by tie** [comment]
I’ve installed macOS on my old Intel MacBook, it’ll take some time to build stdenv (oof dual core CPU), but hopefully I’ll be able to test this PR on x86_64-darwin by the end of this week.

**Comment by tie** [comment]
I can reproduce `test-fs-readv` and `test-fs-readv-sync` test failures on x86_64-darwin (see https://github.com/NixOS/nixpkgs/pull/327653#issuecomment-2251926462) without Nix sandbox. Not sure what the actual issue is, my completely unfounded guess is that it is caused by different SDK versions being used for x86_64-darwin and aarch64-darwin. Unfortunately, currently I don’t have the compute power on MacBook or time to set up Hackintosh to find the root cause.

That said, <em>these tests were failing before changes in this PR</em>, so I’m adding them to the list of disabled tests for x86_64-darwin.

**Comment by emilazy** [comment]
> I’ve installed macOS on my old Intel MacBook, it’ll take some time to build stdenv (oof dual core CPU), but hopefully I’ll be able to test this PR on x86_64-darwin by the end of this week.

Sometimes this isn’t possible due to conflicts, but I would generally recommend basing your `staging` PRs on `master`, so that you can test them without rebuilding the world. Of course, that mostly helps when it’s `staging` because it causes too many rebuilds, rather than because it relates to some actual stdenv change.

Going to try these builds out now (…rebased onto `master`), thanks for enduring with this PR!

File: pkgs/development/web/nodejs/nodejs.nix (hunk 4 above)
Comment (emilazy) [reply pkgs/development/web/nodejs/nodejs.nix:143]: Filed https://github.com/ninja-build/ninja/issues/2475.

File: pkgs/development/web/nodejs/nodejs.nix [hunk 8]
```diff
@@ -140,9 +245,10 @@ let
 
     __darwinAllowLocalNetworking = true; # for tests
 
-    # TODO: what about tests when cross-compiling?
-    # Note that currently stdenv does not run check phase if build ≠ host.
-    doCheck = true;
+    doCheck = canExecute;
+
+    # See https://github.com/nodejs/node/issues/22006
+    enableParallelChecking = false;
```
Comment (emilazy) [inline pkgs/development/web/nodejs/nodejs.nix:251]: It seems like upstream hasn’t seen any issues here for years and closed this issue as a result. I’m wondering if it really is an upstream issue, but if so, it’d be good to ping the issue with example logs to hopefully get it reopened (and maybe even fixed).

**Comment by emilazy** [comment]
Here we go again

Do you want the application “node” to accept incoming network connections? Do you want the application “node” to accept incoming network connections? Do you want the application “node” to accept incoming network connections? Do you want the application “node” to accept

**Comment by emilazy** [comment]
Successfully built `nodejs_{18,20,22}` on `x86_64-darwin` with `sandbox = false`. Doing the slim builds and checking `aarch64-darwin` and then I’ll merge.

**Comment by emilazy** [comment]
Failed on the `aarch64-darwin` community builder with `--option sandbox relaxed`; a lot of errors like this:

```
../../test/cctest/test_inspector_socket.cc:379: Failure
Expected equality of these values:
  0
  err
    Which is: -16

[  FAILED  ] InspectorSocketTest.HostIPv6NonRoutableDual (0 ms)
[ RUN      ] InspectorSocketTest.HostIPv4InSquareBrackets
../../test/cctest/test_inspector_socket.cc:361: Failure
Expected equality of these values:
  0
  uv_listen(reinterpret_cast<uv_stream_t*>(&server), 1, on_new_connection)
    Which is: -48

[-AI] async    0x101bd7060
[R--] tcp      0x101bd6d88
```

The full list:

```
[  FAILED  ] 40 tests, listed below:
[  FAILED  ] InspectorSocketTest.ReadsAndWritesInspectorMessage
[  FAILED  ] InspectorSocketTest.BufferEdgeCases
[  FAILED  ] InspectorSocketTest.AcceptsRequestInSeveralWrites
[  FAILED  ] InspectorSocketTest.ExtraTextBeforeRequest
[  FAILED  ] InspectorSocketTest.RequestWithoutKey
[  FAILED  ] InspectorSocketTest.KillsConnectionOnProtocolViolation
[  FAILED  ] InspectorSocketTest.CanStopReadingFromInspector
[  FAILED  ] InspectorSocketTest.CloseDoesNotNotifyReadCallback
[  FAILED  ] InspectorSocketTest.CloseWorksWithoutReadEnabled
[  FAILED  ] InspectorSocketTest.ReportsHttpGet
[  FAILED  ] InspectorSocketTest.HandshakeCanBeCanceled
[  FAILED  ] InspectorSocketTest.GetThenHandshake
[  FAILED  ] InspectorSocketTest.WriteBeforeHandshake
[  FAILED  ] InspectorSocketTest.CleanupSocketAfterEOF
[  FAILED  ] InspectorSocketTest.EOFBeforeHandshake
[  FAILED  ] InspectorSocketTest.Send1Mb
[  FAILED  ] InspectorSocketTest.ErrorCleansUpTheSocket
[  FAILED  ] InspectorSocketTest.NoCloseResponseFromClient
[  FAILED  ] InspectorSocketTest.HostCheckedForGET
[  FAILED  ] InspectorSocketTest.HostCheckedForUPGRADE
[  FAILED  ] InspectorSocketTest.HostIPChecked
[  FAILED  ] InspectorSocketTest.HostNegativeIPChecked
[  FAILED  ] InspectorSocketTest.HostIpOctetOutOfIntRangeChecked
[  FAILED  ] InspectorSocketTest.HostIpOctetFarOutOfIntRangeChecked
[  FAILED  ] InspectorSocketTest.HostIpEmptyOctetStartChecked
[  FAILED  ] InspectorSocketTest.HostIpEmptyOctetMidChecked
[  FAILED  ] InspectorSocketTest.HostIpEmptyOctetEndChecked
[  FAILED  ] InspectorSocketTest.HostIpTooFewOctetsChecked
[  FAILED  ] InspectorSocketTest.HostIpTooManyOctetsChecked
[  FAILED  ] InspectorSocketTest.HostIpInvalidOctalOctetStartChecked
[  FAILED  ] InspectorSocketTest.HostIpInvalidOctalOctetMidChecked
[  FAILED  ] InspectorSocketTest.HostIpInvalidOctalOctetEndChecked
[  FAILED  ] InspectorSocketTest.HostIpLeadingZeroStartChecked
[  FAILED  ] InspectorSocketTest.HostIpLeadingZeroMidChecked
[  FAILED  ] InspectorSocketTest.HostIpLeadingZeroEndChecked
[  FAILED  ] InspectorSocketTest.HostIPNonRoutable
[  FAILED  ] InspectorSocketTest.HostIPv6NonRoutable
[  FAILED  ] InspectorSocketTest.HostIPv6NonRoutableDual
[  FAILED  ] InspectorSocketTest.HostIPv4InSquareBrackets
[  FAILED  ] InspectorSocketTest.HostIPv6InvalidAbbreviation
```

Running again with the sandbox off.

**Comment by emilazy** [comment]
Uh, actually I think that was with `--option sandbox false`; I got my builds mixed up. Trying with the sandbox *on* instead.

**Comment by emilazy** [comment]
The slim variants build fine on `x86_64-darwin` with sandbox disabled, as expected.

**Comment by emilazy** [comment]
The slim variants build fine on `aarch64-darwin` with `sandbox = relaxed`. Maybe it’s something about tests interfering when run in parallel with the sandbox disabled. Trying with `-j1`.

**Comment by tie** [comment]
>Uh, actually I think that was with `--option sandbox false`; I got my builds mixed up. Trying with the sandbox on instead.

Does Nix actually use `--option sandbox false` from command line? I though this setting has no effect unless set from `nix.conf` (and also needs `launchctk kickstart -k` for Nix daemon).

**Comment by emilazy** [comment]
It does as long as you are in `trusted-users`, which you shouldn’t be (it’s a trivial root escalation vulnerability), but this is on the community builders.

**Comment by emilazy** [comment]
Okay, both Node.js 18 variants went through on `aaarch64-darwin` with the sandbox off and `-j1`. If you think this is ready to merge, I’m satisfied to do so as soon as I confirm that the default Node.js version also builds on there.

I do worry that the potential hint of test flakiness when builds are run in parallel (a common theme with loopback networking tests) will mean further Hydra build failures and result in us having to turn tests off on Darwin again, but I’m willing to chance it. (We could at least just turn those specific tests off if it ends up being necessary rather than the whole suite.)

**Review by emilazy — APPROVED** [review]
It succeeded. This PR has gone on for long enough. Let’s ship it!
